arith/field_hashers/src/poseidon/impls.rs

```diff
@@ -0,0 +1,210 @@
+use std::{
+    fmt::Debug,
+    ops::{Add, AddAssign, Mul, MulAssign},
+};
+
+use arith::{Field, FieldForECC};
+use tiny_keccak::{Hasher, Keccak};
+
+use super::compile_time::compile_time_sbox_alpha;
+use crate::FiatShamirFieldHasher;
+
+pub trait PoseidonStateTrait:
+    Sized
+    + Default
+    + Clone
+    + Debug
+    + Add<Self, Output = Self>
+    + for<'a> Add<&'a Self, Output = Self>
+    + AddAssign<Self>
+    + for<'a> AddAssign<&'a Self>
+    + Mul<Self, Output = Self>
+    + for<'a> Mul<&'a Self, Output = Self>
+    + MulAssign<Self>
+    + for<'a> MulAssign<&'a Self>
+{
+    /// ElemT is the element (base field) type constructing the poseidon state instance.
+    type ElemT: FieldForECC;
+
+    /// SBOX_POW is the lowest exponential pow for poseidon sbox.
+    const SBOX_POW: usize = compile_time_sbox_alpha::<Self::ElemT>();
+
+    /// FULL_ROUNDS in a poseidon permutation.
+    const FULL_ROUNDS: usize;
+
+    /// PARTIAL_ROUNDS in a poseidon permutation.
+    const PARTIAL_ROUNDS: usize;
+
+    /// STATE_WIDTH stands for the number of field elements that build up a poseidon state.
+    const STATE_WIDTH: usize;
+
+    /// CAPACITY is the nubmer of output field elements (field here stands for InputF).
+    ///
+    /// We ensure (roughly) the collision resilience is 128 bits by FIELD_SIZE, i.e.,
+    /// how many bits we need to store a field element.
+    const CAPACITY: usize = 128 / <Self::ElemT as Field>::FIELD_SIZE * 2;
+
+    /// RATE is the number of input elements in a round of sponge absorbing.
+    /// The invariant here is RATE + CAPACITY = STATE_WIDTH
+    const RATE: usize = Self::STATE_WIDTH - Self::CAPACITY;
+
+    /// from_elems constructs an instance of field hasher state from the input field elements.
+    fn from_elems(elems: &[Self::ElemT]) -> Self;
+
+    /// to_elems is a conjugate method of from_elems, that separates the state into elems.
+    fn to_elems(&self) -> Vec<Self::ElemT>;
+
+    /// apply_mds_matrix applies MDS matrix back to the Poseidon state.
+    fn apply_mds_matrix(&mut self, mds_matrix: &[Self]) {
+        let res = mds_matrix
+            .iter()
+            .map(|mds_col| {
+                let hadamard: Self = self.clone() * mds_col;
+                hadamard.to_elems().iter().sum()
+            })
+            .collect::<Vec<_>>();
+
+        *self = Self::from_elems(&res)
+    }
+
+    /// full_round_sbox applies x -> x^(sbox_exp) to all elements in the state.
+    fn full_round_sbox(&mut self) {
+        let mut elts = self.to_elems();
```

Comment:
this is fine for trait implementations. but for instantiations, we prob want to overload it so that we don't need to perform two unnecessary clones

Reply:
good point - I am just leaving a baseline trait impl for functionality, but truly I should leave an override in M31x16 impl side, see latest commit.